Get chaos environments command

shivam-Purohit opened the pull request:
refers issue #164

Addition of new commands :
```
litmusctl get chaos-environment --project-id="" --environment-id=""
litmusctl list chaos-environment --project-id=""
```

shivam-Purohit commented:
Suggest if further changes are required.

shivam-Purohit commented:
@SarthakJain26 @Saranya-jena can I get a review?

SarthakJain26 commented on pkg/cmd/get/environment.go:
```diff
@@ -0,0 +1,89 @@
+		environmentList, err := environment.GetEnvironmentList(projectID, credentials);
+		if err != nil {
+			if strings.Contains(err.Error(), "permission_denied") {
+				utils.Red.Println("❌ The specified Project ID doesn't exist.")
```
Please update the message to, "You don't have enough permissions to access this resource"

SarthakJain26 commented on pkg/cmd/get/environment.go:
```diff
@@ -0,0 +1,89 @@
+func init() { 
+	GetCmd.AddCommand(ChaosEnvironmentCmd);
+	ChaosEnvironmentCmd.Flags().String("project-id", "", "Set the project-id to list Chaos Environment from a particular project.")
+	ChaosEnvironmentCmd.Flags().String("environment-id", "", "Set the environment-id to list Chaos Environment")
```
```suggestion
	ChaosEnvironmentCmd.Flags().String("environment-id", "", "Set the environment-id to get Chaos Environment")
```

SarthakJain26 commented on pkg/cmd/get/environment.go:
```diff
@@ -0,0 +1,89 @@
+}
+func init() { 
+	GetCmd.AddCommand(ChaosEnvironmentCmd);
+	ChaosEnvironmentCmd.Flags().String("project-id", "", "Set the project-id to list Chaos Environment from a particular project.")
```
```suggestion
	ChaosEnvironmentCmd.Flags().String("project-id", "", "Set the project-id to get Chaos Environment from a particular project.")
```

SarthakJain26 commented on pkg/cmd/list/environment.go:
```diff
@@ -0,0 +1,69 @@
+		environmentList, err := environment.GetEnvironmentList(projectID, credentials);
+		if err != nil {
+			if strings.Contains(err.Error(), "permission_denied") {
+				utils.Red.Println("❌ The specified Project ID doesn't exist.")
```
Please update the message to, "You don't have enough permissions to access this resource"

SarthakJain26 commented on pkg/cmd/list/list.go:
```diff
@@ -0,0 +1,31 @@
+var ListCmd = &cobra.Command{
+	Use: "list",
+	Short: `Examples:
+		#get list of chaos Chaos Experiments
```
```suggestion
		#get list of chaos Chaos Environments
```

SarthakJain26 commented:
@shivam-Purohit can you please also share the screen shots of the command and the output

SarthakJain26 commented:
@shivam-Purohit please add these commands in the readme and usage docs as well

shivam-Purohit commented:
> @shivam-Purohit can you please also share the screen shots of the command and the output

sure!
![ubuntuMachine  Running  - Oracle VM VirtualBox 29-12-2023 10_17_17](https://github.com/litmuschaos/litmusctl/assets/91889807/f7931ef0-b73e-451d-98cb-8108c7976dcd)
![ubuntuMachine  Running  - Oracle VM VirtualBox 29-12-2023 10_16_26](https://github.com/litmuschaos/litmusctl/assets/91889807/e5bf5106-05bb-4eeb-970b-a20150f5d2ab)

shivam-Purohit commented:
> @shivam-Purohit please add these commands in the readme and usage docs as well

I added these in 0.23.0. Do we need to add in any other versions? i think we do not need that.

@SarthakJain26 I fixed all the changes you requested. Let me know if further changes are required.

SarthakJain26 commented:
@shivam-Purohit , you have displayed the complete response coming from the API which is not a very good user experience. You will have to extract required fields and you can display them in table format. I have attached the screenshot for reference. You can display environmentID, name, createdAt and createdBy fields. Please update the PR accordingly
<img width="796" alt="Screenshot 2023-12-29 at 11 24 06 AM" src="https://github.com/litmuschaos/litmusctl/assets/45846277/7a9d2fe4-bb0b-46d8-a0ea-7fc24693d658">

shivam-Purohit commented:
hey @SarthakJain26 I am currently getting the response as null in many fields which ig should not. I tried rechecking the API. Is there something I am doing here to retrieve the data?

SarthakJain26 commented:
> hey @SarthakJain26 I am currently getting the response as null in many fields which ig should not. I tried rechecking the API. Is there something I am doing here to retrieve the data?

@shivam-Purohit Possibly yes. You need to check on what is the graphql request you are sending and then what is the data you are expecting back. While decoding the response from the graphql server you need to check if the struct you are using to decode the response has the required fields or not.

shivam-Purohit commented:
@SarthakJain26 the error was actually due to query itself which we were sending through api. I changed it a bit for the newcommand. 
I also refactored the output and reused some other components for similarity and reusability.
![ubuntuMachine  Running  - Oracle VM VirtualBox 07-01-2024 18_21_19](https://github.com/litmuschaos/litmusctl/assets/91889807/4d9f34a0-1944-4ae3-a2ce-14aa100d8271)

SarthakJain26 commented:
Thanks @shivam-Purohit, update output looks awesome

Nageshbansal commented:
@SarthakJain26, do we really need two separate commands for this, i.e., list and get? I think  `get` should be enough only to show a list of all environments, as that is what we're doing for experiments and infra.

SarthakJain26 commented:
> @SarthakJain26, do we need two separate commands for this, i.e., list and get? Maybe `get` should be enough only to show a list of all environments, as that is what we're doing for experiments and infra.

@Nageshbansal , by convention, list command can be used to list the resources under the project with required info, and get command can be used to get additional info about a particular resource. We can follow the same in other places as well. 
Adding to this @shivam-Purohit please let us know if you can add other details also in the get environment command in couple of days so that we can merge this PR and take it in this release.

shivam-Purohit commented:
> > @SarthakJain26, do we need two separate commands for this, i.e., list and get? Maybe `get` should be enough only to show a list of all environments, as that is what we're doing for experiments and infra.
> 
> @Nageshbansal , by convention, list command can be used to list the resources under the project with required info, and get command can be used to get additional info about a particular resource. We can follow the same in other places as well. Adding to this @shivam-Purohit please let us know if you can add other details also in the get environment command in couple of days so that we can merge this PR and take it in this release.


```
type Environment struct {
	ProjectID     string          `json:"projectID"`
	EnvironmentID string          `json:"environmentID"`
	Name          string          `json:"name"`
	Description   *string         `json:"description"`
	Tags          []string        `json:"tags"`
	Type          EnvironmentType `json:"type"`
	CreatedAt     string          `json:"createdAt"`
	CreatedBy     *UserDetails    `json:"createdBy"`
	UpdatedBy     *UserDetails    `json:"updatedBy"`
	UpdatedAt     string          `json:"updatedAt"`
	IsRemoved     *bool           `json:"isRemoved"`
	InfraIDs      []string        `json:"infraIDs"`
}
```

these are the details we can get from the response. @SarthakJain26 let me know what details you want to add, imo we can add type, description or maybe updated at and updated by.

shivam-Purohit commented:
> @shivam-Purohit you can also add the InfraIDs along with what you mentioned so that it infras within the environment can be linked.

sure!

SarthakJain26 commented:
@shivam-Purohit can you please fix the fmt issues due to which build pipeline is failing and also sign your commits so that DCO can pass. We can then merge the PR

shivam-Purohit commented:
> can you please fix the fmt issues due to which build pipeline is failing and also sign your commits so that DCO can pass. We can then merge the PR

should I squash the commits into one?

Nageshbansal commented:
@shivam-Purohit Squashing the commit is not required ( it will be taken care of while merging) , for now please fix the fmt in go files ( you can use gofmt command to check which files are not correctly formatted)

Nageshbansal commented:
@shivam-Purohit Squashing the commit is not required ( it will be taken care of while merging) , for now please fix the fmt in go files ( you can use gofmt command to check which files are not correctly formatted)

Nageshbansal commented:
For DCO, you can refer to : https://github.com/litmuschaos/litmus/blob/master/CONTRIBUTING.md#sign-your-work-with-developer-certificate-of-origin

shivam-Purohit commented:
@Nageshbansal my first commit was unsigned i could not find any other way to sign it. So I merged it. If the workflows get approval i can test the changes i just made . I use gofmt -w . that should be able to get rid of the fmt error

shivam-Purohit commented:
> @shivam-Purohit can you please fix the fmt issues due to which build pipeline is failing and also sign your commits so that DCO can pass. We can then merge the PR

@SarthakJain26 can you run the workflow again ig this should be fine now at least the fmt error

Jonsy13 commented on pkg/cmd/get/environment.go:
```diff
@@ -0,0 +1,114 @@
+			if environmentListData[i].EnvironmentID == environmentID {
+				intUpdateTime, err := strconv.ParseInt(environmentListData[i].UpdatedAt, 10, 64)
+				if err != nil {
+					fmt.Println("Error converting UpdatedAt to int64:", err)
```
utils.Red.Println

Jonsy13 commented on pkg/cmd/get/environment.go:
```diff
@@ -0,0 +1,114 @@
+			utils.White_B.Print("\nEnter the Project ID: ")
+			fmt.Scanln(&projectID)
+
+			for projectID == "" {
```
Can we use `if` here?

shivam-Purohit replied:
I picked it from experiments get command so the same code is repeated there but doesn't make any sense to put for and then exit anyway. If would be a better choice

Jonsy13 commented on pkg/cmd/get/environment.go:
```diff
@@ -0,0 +1,114 @@
+			utils.White_B.Print("\nEnter the Environment ID: ")
+			fmt.Scanln(&environmentID)
+
+			for environmentID == "" {
```
Can we use if here?

Jonsy13 commented on pkg/cmd/get/environment.go:
```diff
@@ -0,0 +1,114 @@
+				updatedTime := time.Unix(intUpdateTime, 0).String()
+				intCreatedTime, err := strconv.ParseInt(environmentListData[i].CreatedAt, 10, 64)
+				if err != nil {
+					fmt.Println("Error converting CreatedAt to int64:", err)
```
utils.Red.Println

shivam-Purohit commented:
@Jonsy13 @Saranya-jena I made the changes suggested you can take a look. Suggest if any further changes need to be done.

SarthakJain26 commented:
@shivam-Purohit please fix the build pipeline

shivam-Purohit commented:
@SarthakJain26 fixed the fmt error. Hope it resolves the build workflow

shivam-Purohit commented:
Looks like it worked!
